Added timezone handling

peterfabian opened the pull request:
Fixes #489, fixes #1414.

This PR changes semantics of `date_created` columns so that they are no longer storing UTC time, but local time. 
That makes wc-admin API consistent with WordPress REST API, where `after` and `before` parameters are compared against `post_date`, not `post_date_gmt`.

It also allows us to slice and dice the data into time intervals in the database without requiring any additional database commands. If we stored the UTC timestamp and wanted to group the data by time intervals, we would have to convert the timestamp to correct timezone in the database, which would require having timezone tables filled in the db (by running `mysql_tzinfo_to_sql /usr/share/zoneinfo | mysql -u root mysql`) and that might be difficult on some hostings. I wanted to avoid that.

That also means that datetime without timezone info is assumed to be in local timezone set in WordPress general settings, i.e. `after=2019-02-01T00:00:00` means after Feb 1st in the local timezone, not UTC.

Also, this PR adds conversion of forced timezone information to local timezone, i.e. `2019-02-01T00:00:00+00:00` is a UTC time, which gets converted to `2019-02-01 01:00:00` in Berlin or `2019-01-31 19:00:00` in New York for db queries.

### Detailed test instructions:

- Database rebuild is required for this change, as previous data are no longer valid, plus data type of `date_created` column changed from timestamp to datetime
- check that correct orders are being included when running with different date/time parameters

<!--- Note: When displaying information based on sample data, such as SwaggerHub, 
be sure to detail parts affected in Release Notes --->

peterfabian committed:
Changed the date_created columns to datetime.

date_created should store local date and time, not UTC timestamp, as the data needs to be sliced and diced by local date/time, not by UTC date/time.

peterfabian committed:
Store datetimes in local time instead of UTC.

peterfabian committed:
Normalize datetimes before doing any further processing with the input from the user.

Also, passing DateTime objects around instead of string so that timezone info can be easily attached to it.

peterfabian committed:
Removed obsolete comment.

peterfabian committed:
Corrected output of GMT time.

peterfabian committed:
Centralized default before and after parameters.

peterfabian committed:
Fixed number of parameters to the function.

peterfabian committed:
Fixed parameter name from previous copy paste error.

peterfabian committed:
Removed superfluous date setting.

peterfabian committed:
Use default after and before parameter values if no values are provided by the user.

peterfabian committed:
Accept before and after parameters as DateTime or string.

peterfabian committed:
Merge branch 'master' into fix/489-date_date_gmt

justinshreve commented:
I'll do a deeper review of this tomorrow, but I did some testing on this today when looking at #1414 and it fixes the issue there 👍. I edited your PR description so it will get closed when this merges.

justinshreve commented on includes/class-wc-admin-reports-interval.php:
```diff
@@ -26,6 +26,44 @@ class WC_Admin_Reports_Interval {
+	 *
+	 * @return DateTime
+	 */
+	public static function default_before() {
```
Nice idea/refactor here, allowing the different data stores to be cleaned up when setting defaults 👍.

justinshreve commented on includes/class-wc-admin-reports-interval.php:
```diff
@@ -422,9 +461,6 @@ public static function next_year_start( $datetime, $reversed = false ) {
-		// $result_datetime           =
-		// $result_timestamp_adjusted = $result_datetime->format( 'U' ) - 1;
-		// $result_datetime->setTimestamp( $result_timestamp_adjusted );
 		return call_user_func( array( __CLASS__, "next_{$time_interval}_start" ), $datetime, $reversed );
```
Since it looks like `$time_interval` comes from the `query` args, should we do any checking to be safe that it is a valid interval function before calling it?

peterfabian replied:
I think it's validated for REST queries via the controllers, e.g. 

https://github.com/woocommerce/wc-admin/blob/368c02e53086f9f3a21451e14b2b08ab664fbd15/includes/api/class-wc-admin-rest-reports-orders-stats-controller.php#L366-L379

However, if used directly as `WC_Admin_Reports_*_Query` or `WC_Admin_Reports_*_Data_Store`, then invalid value might result in fatals, so it's a valid concern. I think there are more args where this is true.

justinshreve replied:
Ah yeah, good point on REST handing validation.

It might be worth adding some protection to this and other args later, but definitely doesn't need to be a blocker. 👍

emdashcodes reviewed (approved):
I have one comment/question, but otherwise this is looking good to me and is testing great after rebuilding. Hour charts are working well and displaying the correct order time. Other intervals are also looking good. All tests pass via `phpunit`. Pre-approving.

timmyc commented:
Looks all good to go after a rebase, thanks @peterfabian for getting this one fixed up!

peterfabian committed:
Merge branch 'master' into fix/489-date_date_gmt

# Conflicts:
#	includes/class-wc-admin-api-init.php

peterfabian committed:
Changed the date_created columns to datetime, in the new location of db create script.

date_created should store local date and time, not UTC timestamp, as the data needs to be sliced and diced by local date/time, not by UTC date/time.

peterfabian commented:
Thanks for the reviews and testing.
